### ISSUE TITLE
Float/Double Support for Formatter

### ISSUE THREAD

**Issue by MeFisto94** [konsoletyper/teavm#472]
Currently, something like `String.format("%f", 1.0f)` leads to TFormatter#formatValue:
`throw new UnknownFormatConversionException(String.valueOf(specifier));`

**Comment by konsoletyper** [comment]
Right, `String.format` has only limited support, and I have no plans to support it fully.

**Comment by ScraM-Team** [comment]
@MeFisto94 , I ran into this recently too.  I needed just 1 decimal place so I found the fractional part via Math.floor(doubleValue * 10.0d) % 10.

**Comment by tommyettinger** [comment]
Is there any interest in a PR for this? My library [formic](https://github.com/tommyettinger/formic) currently works on GWT by using a JS library, [format4js](https://github.com/madogiwa/format4js), but I don't know how to do something like that when it's used by TeaVM. Because formic knows it isn't running on GWT, it thinks it's running on a platform with complete `String.format()`, and so calling formic's `Stringf.format()` falls back to normal `String.format()`. I recently worked with some code that converts doubles and floats to Strings at a low level in pure Java. I used [Ryu](https://github.com/ulfjack/ryu), which has some `printf()` support for C but not Java. It wouldn't help with dates or times, but it could handle variable lengths of precision, I think, with some adjustment. [My code that modifies Ryu is here.](https://github.com/tommyettinger/digital/blob/main/src/main/java/com/github/tommyettinger/digital/RyuDouble.java)

**Comment by konsoletyper** [comment]
@tommyettinger I don't think TeaVM should use format4js for this purpose. If you want to help with it, you can contribute your code to [Formatter](https://github.com/konsoletyper/teavm/blob/c9d97b7144a4689d0044a1f07193c6a8e01aa7db/classlib/src/main/java/org/teavm/classlib/java/util/TFormatter.java) implementation.

**Comment by tommyettinger** [comment]
I'm certainly interested, and yeah I would probably use Formatter (I was just looking there!). My concern is that it would probably need a fairly substantial amount of code from Ryu, which I believe is compatibly licensed (Apache 2.0). I could take RyuFloat and RyuDouble in as package-private for java.util, which might actually be what the JDK does in recent versions; I know they changed their float-printing algorithm recently. I could also use other number-printing code if there's some here.

Related to that last point, I noticed that `String.valueOf(Integer.MIN_VALUE)` and `String.valueOf(Long.MIN_VALUE)` both return `"-0"` instead of a large negative number.

**Comment by konsoletyper** [comment]
@tommyettinger you don't need code for formatting numbers, dates, etc, it's already there. See [java.text](https://github.com/konsoletyper/teavm/tree/c9d97b7144a4689d0044a1f07193c6a8e01aa7db/classlib/src/main/java/org/teavm/classlib/java/text) implementation with all formatters available. The only thing necessary is to properly parse format spec and call corresponding code. If you want to create a PR, please don't forget to add [tests](https://github.com/konsoletyper/teavm/blob/c9d97b7144a4689d0044a1f07193c6a8e01aa7db/tests/src/test/java/org/teavm/classlib/java/util/FormatterTest.java) and run them properly (`./gradlew test`). Also, don't forget to run `./gradlew check` to check if code follows conventions.

**Comment by tommyettinger** [comment]
Very nice, I'm glad there's so much already here! I'll look around java.text and see which additional formatting modes can be supported right away, and which ones might be less immediate (`%A` is probably less important).
